assets/js/payment-methods-demo/stripe-utils/index.js

```diff
@@ -151,11 +153,106 @@ const canDoApplePay = ( paymentRequest ) => {
 	} );
 };
 
+const isNonFriendlyError = ( type ) =>
+	[
+		errorTypes.INVALID_REQUEST,
+		errorTypes.API_CONNECTION,
+		errorTypes.API_ERROR,
+		errorTypes.AUTHENTICATION_ERROR,
+		errorTypes.RATE_LIMIT_ERROR,
+	].includes( type );
+
+const getErrorMessageForCode = ( code ) => {
+	const messages = {
+		[ errorCodes.INVALID_NUMBER ]: __(
+			'The card number is not a valid credit card number.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INVALID_EXPIRY_MONTH ]: __(
+			"The card's expiration month is invalid.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INVALID_EXPIRY_YEAR ]: __(
+			"The card's expiration year is invalid.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INVALID_CVC ]: __(
+			"The card's security code is invalid.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCORRECT_NUMBER ]: __(
+			'The card number is incorrect.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCOMPLETE_NUMBER ]: __(
+			'The card number is incomplete.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCOMPLETE_CVC ]: __(
+			"The card's security code is incomplete.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCOMPLETE_EXPIRY ]: __(
+			"The card's expiration date is incomplete.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.EXPIRED_CARD ]: __(
+			'The card has expired.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCORRECT_CVC ]: __(
+			"The card's security code is incorrect.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INCORRECT_ZIP ]: __(
+			"The card's zip code failed validation.",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.INVALID_EXPIRY_YEAR_PAST ]: __(
+			"The card's expiration year is in the past",
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.CARD_DECLINED ]: __(
+			'The card was declined.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.MISSING ]: __(
+			'There is no card on a customer that is being charged.',
+			'woocommerce-gateway-stripe'
+		),
+		[ errorCodes.PROCESSING_ERROR ]: __(
+			'An error occurred while processing the card.',
+			'woocommerce-gateway-stripe'
+		),
+	};
+	return messages[ code ] || '';
+};
+
+const getErrorMessageForTypeAndCode = ( type, code = '' ) => {
+	switch ( type ) {
+		case errorTypes.INVALID_EMAIL:
+			return __(
+				'Invalid email address, please correct and try again.',
```

Comment:
Are gateways validating checkout fields rather than checkout itself?

Reply:
This was pulled directly from the stripe payment gateway and I _think_ there are some stripe payment method variations the stripe form has an option for an email field (which is why they have it). In our specific implementation this code would probably never get fired (b/c we're not using that variation) but I just copied all codes in.

Reply:
The other possibility is that this may be present because at one point Woo Core didn't have great email field validation and sometimes the gateway would reject the email submitted with the payment? IF this error gets returned when trying to create the stripe source, I can still have checkout handle invalidating the email field during process (with a message for why it was invalidated).